src/rewrites.rs

```diff
@@ -588,3 +712,274 @@ fn check_pat(
+            .iter()
+            .map(|x| x.search(&runner.egraph))
+            .collect();
+
+        // For each multi rule
+        for (i, rule) in self.rules.iter().enumerate() {
+            let map_1 = &self.src_pat_maps[i].0;
+            let map_2 = &self.src_pat_maps[i].1;
+            let matches_1 = &matches[map_1.index];
+            let matches_2 = &matches[map_2.index];
+            for match_1 in matches_1 {
+                for match_2 in matches_2 {
+                    if match_1.eclass == match_2.eclass {
+                        // We don't want to apply multi-pattern rules on the same eclass
+                        continue;
+                    }
+                    for subst_1 in &match_1.substs {
+                        for subst_2 in &match_2.substs {
+                            // De-canonicalize the substitutions
+                            let subst_1_dec = decanonicalize(subst_1, &map_1.var_map);
+                            let subst_2_dec = decanonicalize(subst_2, &map_2.var_map);
+                            // Check if two substitutions have matching shared variables
+                            if compatible(&subst_1_dec, &subst_2_dec, &map_1.var_map) {
+                                // If so, merge two substitutions
+                                let merged_subst =
+                                    merge_subst(subst_1_dec, subst_2_dec, &map_1.var_map);
+
+                                // check_pat on both dst patterns
+                                if check_pat(rule.2.ast.as_ref(), &mut runner.egraph, &merged_subst)
+                                    .0
+                                {
+                                    if check_pat(
+                                        rule.3.ast.as_ref(),
+                                        &mut runner.egraph,
+                                        &merged_subst,
+                                    )
+                                    .0
+                                    {
+                                        // apply dst patterns, union
+                                        let id_1 = rule.2.apply_one(
+                                            &mut runner.egraph,
+                                            match_1.eclass,
+                                            &merged_subst,
+                                        )[0];
+                                        runner.egraph.union(id_1, match_1.eclass);
+                                        let id_2 = rule.3.apply_one(
+                                            &mut runner.egraph,
+                                            match_2.eclass,
+                                            &merged_subst,
+                                        )[0];
+                                        runner.egraph.union(id_2, match_2.eclass);
+                                    }
+                                }
+                            }
+                        }
```

Comment:
Move to a separate method to reduce indentation.

Reply:
Done